### PR TITLE
Add post on generic API support unconventional array indices

### PR DESCRIPTION
**Pull request by timholy**
A companion to https://github.com/JuliaLang/julia/pull/16260. Probably the easiest way for folks to learn the new API.

### CODE REVIEW

File: blog/_posts/2016-06-12-offset-arrays.md [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+this storage to match the indices of some input, these will fail for
+inputs with unconventional indices.
+
+The generic replacement for such patterns is a new function,
+`allocate_for(storagetype, referencearray, [shape])`.  `storagetype`
+indicates the kind of underlying "conventional" behavior you'd like,
+e.g., `Array{Int}` or `BitArray`. `referencearray` is the array which
+might have unconventional indexing, i.e., what you'd like to match
+(both in type and indices). Optionally you can supply particular
+`shape` inforation as the third argument.
```
Comment (mauro3) [inline blog/_posts/2016-06-12-offset-arrays.md]: information

File: blog/_posts/2016-06-12-offset-arrays.md [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+unavoidable: earlier versions of Julia defined `first(::Colon) = 1`,
+meaning that the first index along a dimension indexed by `:` is 1.
+This definition can no longer be justified, so it was deprecated.
+
+# Creating new "offset" array types
+
+If you're creating a new `AbstractArray` type and want it to support
+unconventional indices, there are a few other things you need to know.
+A good model to follow is `test/offsetarray.jl` in Julia's source
+directory.
```
Comment (mauro3) [inline blog/_posts/2016-06-12-offset-arrays.md]: [`test/offsetarray.jl`](https://github.com/JuliaLang/julia/tree/master/test/offsetarray.jl)

File: blog/_posts/2016-06-12-offset-arrays.md [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+finally select a single one to represent the "template" type.  For
+example, `test/offsetarray.jl` contains the definition
+
+```jl
+Base.promote_indices(a::OffsetArray, b::OffsetArray) = a
+```
+
+which ends up meaning that any list of `OffsetArray`s will result in a
+call with `referencearray` of type `OffsetArray`.  The absence of any
+other definitions implies that mixutures of conventional and
```
Comment (rfourquet) [inline blog/_posts/2016-06-12-offset-arrays.md]: mixtures!

File: blog/_posts/2016-06-12-offset-arrays.md [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+work. While the "unconventional" array types will be provided via
+packages, the core language has been substantially (though surely
+incompletely) generalized to ensure that algorithms work as expected
+with such arrays.
+
+The purpose of the blog post is to address the question, "what do I
+have to do to support such arrays in my own code?"  First, let's
+address the simplest case: if you know that your code will never need
+to handle arrays with unconventional indexing, hopefully the answer is
+"nothing." Old code, on conventional arrays, should function
```
Comment (pabloferz) [inline blog/_posts/2016-06-12-offset-arrays.md]: `s/"nothing\."/"nothing"\./`

Comment (kmsquire) [inline blog/_posts/2016-06-12-offset-arrays.md]: In American English, periods are always included inside the quote.  In British English, they are always outside.
http://www.thepunctuationguide.com/british-versus-american-style.html, last paragraph

File: blog/_posts/2016-06-12-offset-arrays.md [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+
+Let's walk through a couple of explicit examples. First, if `A` has
+conventional indices, `allocate_for(Array{Int}, A)` would end up
+calling `Array{Int}(size(A))`, and thus return an array.  If `A` is an
+`AbstractArray` type with unconventional indexing, then
+`allocate_for(Array{Int}, A)` should return something that "behaves
+like" an `Array{Int}` but with a shape (including indices) that
+matches `A`.  (The most obvious implementation is to allocate an
+`Array{Int}(size(A))` and then "wrap" it in a type matching `A` that
+shifts the indices.  However, this behavior is not guaranteed.)
```
Comment (tkelman) [inline blog/_posts/2016-06-12-offset-arrays.md]: what does "this behavior is not guaranteed" mean?

**Comment by timholy** [comment]
Moved to http://docs.julialang.org/en/latest/devdocs/offset-arrays/

**Comment by ViralBShah** [comment]
I would have prreferred to publish this as a blog post too. Any reason not to?

**Comment by timholy** [comment]
It's outdated and a subset of what's in devdocs.

I could put a copy of the devdocs page up on the blog, but I wonder about the value of copies. In the list of blog posts, can we just make a hyperlink to the devdocs page?

**Comment by ViralBShah** [comment]
Yes, we could do a short blog post with links to details in the devdocs. The only reason is that the blog gets wider distribution and readership and this is important and exciting enough in my opinion for that.

**Comment by tkelman** [comment]
And blog posts reach exactly the crowd that complains most widely about this :)

**Pull request closed without merging**